`linux/lustre/lustre_user.h` is not available, and breaks configure

Hi,
I downloaded and built 3.2.0rc1.
I run configure with `--with-lustre`, and it breaks with:
```
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking linux/lustre/lustre_user.h usability... no
checking linux/lustre/lustre_user.h presence... no
checking for linux/lustre/lustre_user.h... no
configure: error: in `/dev/shm/ior-3.2.0rc1':
configure: error: --with-lustre was given, <lustre/lustre_user.h> not found
See `config.log' for more details
```

In `config.log` there is:
```
configure:5418: checking for _FILE_OFFSET_BITS value needed for large files
configure:5443: /opt/cluster/openmpi/3.1.2/gnu/7.2.0/bin/mpicc -c -g -O2  conftest.c >&5
configure:5443: $? = 0
configure:5475: result: no
configure:5575: checking linux/lustre/lustre_user.h usability
configure:5575: /opt/cluster/openmpi/3.1.2/gnu/7.2.0/bin/mpicc -c -g -O2  conftest.c >&5
conftest.c:86:10: fatal error: linux/lustre/lustre_user.h: No such file or directory
 #include <linux/lustre/lustre_user.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
configure:5575: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "ior"
| #define PACKAGE_TARNAME "ior"
| #define PACKAGE_VERSION "3.2.0rc1"
...
| #ifdef HAVE_STRING_H
| # if !defined STDC_HEADERS && defined HAVE_MEMORY_H
| #  include <memory.h>
| # endif
| # include <string.h>
| #endif
| #ifdef HAVE_STRINGS_H
| # include <strings.h>
| #endif
| #ifdef HAVE_INTTYPES_H
| # include <inttypes.h>
| #endif
| #ifdef HAVE_STDINT_H
| # include <stdint.h>
| #endif
| #ifdef HAVE_UNISTD_H
| # include <unistd.h>
| #endif
| #include <linux/lustre/lustre_user.h>
configure:5575: result: no
configure:5575: checking linux/lustre/lustre_user.h presence
configure:5575: /opt/cluster/openmpi/3.1.2/gnu/7.2.0/bin/mpicc -E  conftest.c
conftest.c:53:10: fatal error: linux/lustre/lustre_user.h: No such file or directory
 #include <linux/lustre/lustre_user.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
configure:5575: $? = 1
configure: failed program was:
...
```

```console
# ll /usr/include/lustre/lustre_user.h
-rw-r--r-- 1 root root 50K Oct  4 10:28 /usr/include/lustre/lustre_user.h
```

Removing that include from the configure solved my issue:
```console
# grep -rn 'linux/lustre/lustr' .
./configure:5572:        for ac_header in linux/lustre/lustre_user.h lustre/lustre_user.h

# sed -i 's|ac_header in linux/lustre/lustre_user.h|ac_header in|' configure
```

Maybe it should be changed to a XOR with `lustre/lustre_user.h`?

I'd agree to remove "linux/" and replace it with "lustre/lustre_user.h".
Many deployments use it that way.
I added Andreas Dilger on the recipients who can give a more qualified
response.

Am Fr., 14. Dez. 2018 um 13:28 Uhr schrieb Alberto Chiusole <
notifications@github.com>:

> Hi,
> I downloaded and built 3.2.0rc1.
> I run configure with --with-lustre, and it breaks with:
>
> checking for special C compiler options needed for large files... no
> checking for _FILE_OFFSET_BITS value needed for large files... no
> checking linux/lustre/lustre_user.h usability... no
> checking linux/lustre/lustre_user.h presence... no
> checking for linux/lustre/lustre_user.h... no
> configure: error: in `/dev/shm/ior-3.2.0rc1':
> configure: error: --with-lustre was given, <lustre/lustre_user.h> not found
> See `config.log' for more details
>
> In config.log there is:
>
> configure:5418: checking for _FILE_OFFSET_BITS value needed for large files
> configure:5443: /opt/cluster/openmpi/3.1.2/gnu/7.2.0/bin/mpicc -c -g -O2  conftest.c >&5
> configure:5443: $? = 0
> configure:5475: result: no
> configure:5575: checking linux/lustre/lustre_user.h usability
> configure:5575: /opt/cluster/openmpi/3.1.2/gnu/7.2.0/bin/mpicc -c -g -O2  conftest.c >&5
> conftest.c:86:10: fatal error: linux/lustre/lustre_user.h: No such file or directory
>  #include <linux/lustre/lustre_user.h>
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> configure:5575: $? = 1
> configure: failed program was:
> | /* confdefs.h */
> | #define PACKAGE_NAME "ior"
> | #define PACKAGE_TARNAME "ior"
> | #define PACKAGE_VERSION "3.2.0rc1"
> ...
> | #ifdef HAVE_STRING_H
> | # if !defined STDC_HEADERS && defined HAVE_MEMORY_H
> | #  include <memory.h>
> | # endif
> | # include <string.h>
> | #endif
> | #ifdef HAVE_STRINGS_H
> | # include <strings.h>
> | #endif
> | #ifdef HAVE_INTTYPES_H
> | # include <inttypes.h>
> | #endif
> | #ifdef HAVE_STDINT_H
> | # include <stdint.h>
> | #endif
> | #ifdef HAVE_UNISTD_H
> | # include <unistd.h>
> | #endif
> | #include <linux/lustre/lustre_user.h>
> configure:5575: result: no
> configure:5575: checking linux/lustre/lustre_user.h presence
> configure:5575: /opt/cluster/openmpi/3.1.2/gnu/7.2.0/bin/mpicc -E  conftest.c
> conftest.c:53:10: fatal error: linux/lustre/lustre_user.h: No such file or directory
>  #include <linux/lustre/lustre_user.h>
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> configure:5575: $? = 1
> configure: failed program was:
> ...
>
> # ll /usr/include/lustre/lustre_user.h-rw-r--r-- 1 root root 50K Oct  4 10:28 /usr/include/lustre/lustre_user.h
>
> Removing that include from the configure solved my issue:
>
> # grep -rn 'linux/lustre/lustr' ../configure:5572:        for ac_header in linux/lustre/lustre_user.h lustre/lustre_user.h
>
> # sed -i 's|ac_header in linux/lustre/lustre_user.h|ac_header in|' configure
>
> Maybe it should be changed to a XOR with lustre/lustre_user.h?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hpc/ior/issues/115>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE1uyjW5VAMc_o7mUW1ZtZLYPns1OzE1ks5u46dtgaJpZM4ZTiBs>
> .
>


-- 
Dr. Julian Kunkel
Lecturer, Department of Computer Science
+44 (0) 118 378 8218
http://www.cs.reading.ac.uk/
https://hps.vi4io.org/
PGP Fingerprint: 1468 1A86 A908 D77E B40F 45D6 2B15 73A5 9D39 A28E

--- a/configure.ac
+++ b/configure.ac
@@ -34,7 +34,7 @@ esac
 
 # We can't do anything without a working MPI
 AX_PROG_CC_MPI(,,[
-    AC_MSG_FAILURE([MPI compiler requested, but couldn't use MPI.])
+    AC_MSG_FAILURE([MPI compiler requested, but could not use MPI.])
 ])
 
 AC_PROG_RANLIB
@@ -85,7 +85,9 @@ AC_ARG_WITH([lustre],
         [], [with_lustre=check])
 AS_IF([test "x$with_lustre" != xno], [
         AC_CHECK_HEADERS([linux/lustre/lustre_user.h lustre/lustre_user.h], break, [
-                if test "x$with_lustre" != xcheck; then
+                if test "x$with_lustre" != xcheck -a \
+                        "x$ac_cv_header_linux_lustre_lustre_user_h" = "xno" -a \
+                        "x$ac_cv_header_lustre_lustre_user_h" = "xno" ; then
                         AC_MSG_FAILURE([--with-lustre was given, <lustre/lustre_user.h> not found])
                 fi
         ])

--- a/src/aiori-POSIX.c
+++ b/src/aiori-POSIX.c
@@ -32,11 +32,10 @@
 #include <assert.h>
 
 
-#ifdef HAVE_LUSTRE_LUSTRE_USER_H
-#  include <lustre/lustre_user.h>
-#endif
 #ifdef HAVE_LINUX_LUSTRE_LUSTRE_USER_H
 #  include <linux/lustre/lustre_user.h>
+#elif defined(HAVE_LUSTRE_LUSTRE_USER_H)
+#  include <lustre/lustre_user.h>
 #endif
 #ifdef HAVE_GPFS_H
 #  include <gpfs.h>
@@ -282,6 +281,12 @@ void *POSIX_Create(char *testFileName, IOR_param_t * param)
           return 0;
 
 #ifdef HAVE_LUSTRE_LUSTRE_USER_H
+/* Add a #define for FASYNC if not available, as it forms part of
+ * the Lustre O_LOV_DELAY_CREATE definition. */
+#ifndef FASYNC
+#define FASYNC          00020000   /* fcntl, for BSD compatibility */
+#endif
+
         if (param->lustre_set_striping) {
                 /* In the single-shared-file case, task 0 has to creat the
                    file with the Lustre striping options before any other processes
@@ -302,7 +307,8 @@ void *POSIX_Create(char *testFileName, IOR_param_t * param)
                         opts.lmm_stripe_count = param->lustre_stripe_count;
 
                         /* File needs to be opened O_EXCL because we cannot set
-                           Lustre striping information on a pre-existing file. */
+                         * Lustre striping information on a pre-existing file.*/
+
                         fd_oflag |=
                             O_CREAT | O_EXCL | O_RDWR | O_LOV_DELAY_CREATE;
                         *fd = open64(testFileName, fd_oflag, 0664);

